File: node-service/validator.js
var crypto = require('crypto');

module.exports = validate;

var KEY = process.env.KEY;
var SESSION_TIME = 10 * 60;

if (! KEY)
  throw new Error('Key is not defined');

function validate(token) {
  var parts = token.split(' ');

  if (parts.length != 3)
    return false;

  var email = parts[0];
  var ts = parseInt(parts[1]);
  var mac = parts[2];

  if (isNaN(ts) || Date.now()/1000 - ts > SESSION_TIME)
    return false;

  var text = email + ' ' + ts;
  var signed = sign(KEY, text);

  return signed === mac;
}

function sign(key, msg) {
  return crypto.createHmac('sha256', key).update(msg).digest('base64');
}

File: ruby-service/validator.rb
require './signer'

KEY = ENV['KEY']
raise 'No key defined' unless KEY

SESSION_TIME = 10 * 60

module Validator
  def self.valid(token)
    email, ts, mac = token.split ' '
    return false unless email && ts && mac

    age_in_seconds = Time.now.to_i - ts.to_i;
    return false unless age_in_seconds < SESSION_TIME

    text = "#{email} #{ts}"
    signed = Signer.sign KEY, text
    return false unless mac == signed

    true
  end
end

